### PR TITLE
Improvement: add source user and type to tasks and deployments

### LINKED ISSUE #3144: Add author name to pushes
At least for github pushes I believe https://github.com/uselagoon/lagoon/compare/main...chx:patch-1 is all you need.

I could expand on this of course if there's interest.

### PR DESCRIPTION
**Pull request by shreddedbacon**
# General Checklist

- [ ] Affected Issues have been mentioned in the Closing issues section
- [ ] Documentation has been written/updated
- [ ] PR title is ready for inclusion in changelog

# Database Migrations

- [x] If your PR contains a database migation, it **MUST** be the latest in date order alphabetically

Track the user or system that triggered a task or deployment.
It adds `sourceType` and `sourceUser` to Task and Deployments which can only be set when the tasks are created initially. These fields cannot be modified afterwards (beyond direct database interactions)

The two trigger types are `API` and `WEBHOOK`. Deployments currently support both source types.
Tasks are not triggered by webhooks, so only the `API` type is available to it.

The webhook source type will also try and extract information about the user that initiated the webhook call if possible, otherwise the trigger user will just be `webhook`.

API source types will try and retrieve the user email address from the token, if the user triggering the task is a system token, it will extract the `iss` field from the system token, this could result in strange names if a token is really old.

This does not backport the data to existing tasks/deployments as this information is unknown, so the data will be `null` for old tasks and deployments. The thought of adding an `UNKNOWN` type with it being the default, so historical fields would be `UNKNOWN` type. The source user could also default to `unknown` if not provided. Rather than being a `null` field.

# Closing issues

closes #3144 
partially #1980 (better audit capability in lagoon needed generally)

### CODE REVIEW

File: services/api/database/migrations/20240115000000_taskdeployment_source.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.up = async function(knex) {
+    return knex.schema
+    .alterTable('task', function (table) {
+        table.enu('source_type', ['api']);
```
Comment (tobybellwood) [inline services/api/database/migrations/20240115000000_taskdeployment_source.js:8]: acknowledging that there may be future different ways to trigger a task, to be handled in future ENUM migrations

**Comment by shreddedbacon** [comment]
I've added some basic tests to the test suite to check that the fields are populated when deployments are triggered via webhook and api calls both git push and pullrequest, deployenvironmentlatest, and deployenvironment branch.

**Comment by shreddedbacon** [comment]
the `promote` test still seems to use a `test-suite` legacy token, when other tests uses a `ci` user token, this can be resolved in a separate PR though

**Review by tobybellwood — APPROVED** [review]
This looks good to me, and +100 for the tests.

Happy to follow up the errant test token in a follow up.

**Pull request merged**